cluster/images/etcd/rollback/rollback.go

```diff
@@ -0,0 +1,318 @@
+
+		// This is compact key.
+		if !strings.HasPrefix(string(kv.Key), "/") {
+			return nil
+		}
+
+		ttlOpt := store.TTLOptionSet{}
+		if kv.Lease != 0 {
+			ttlOpt = store.TTLOptionSet{ExpireTime: expireTime}
+		}
+
+		if !isTombstone(k) {
+			_, err := st.Set(path.Join("1", string(kv.Key)), false, string(kv.Value), ttlOpt)
+			if err != nil {
+				return err
+			}
+		} else {
+			st.Delete(string(kv.Key), false, false)
+		}
+
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	tx.Unlock()
+
+	traverseAndDeleteEmptyDir(st, "/")
+
+	metadata, hardstate, oldSt := rebuild(migrateDatadir)
+
+	// In the following, it's low level logic that saves metadata and data into v2 snapshot.
+
+	if err := os.RemoveAll(migrateDatadir); err != nil {
+		panic(err)
+	}
+	if err := os.MkdirAll(path.Join(migrateDatadir, "member", "snap"), 0700); err != nil {
+		panic(err)
+	}
+	walDir := path.Join(migrateDatadir, "member", "wal")
+
+	w, err := wal.Create(walDir, metadata)
+	if err != nil {
+		panic(err)
+	}
+	err = w.SaveSnapshot(walpb.Snapshot{Index: hardstate.Commit, Term: hardstate.Term})
+	if err != nil {
+		panic(err)
+	}
+	w.Close()
+
+	nodeIDs := []uint64{}
+	event, err := oldSt.Get(etcdserver.StoreClusterPrefix, true, false)
+	if err != nil {
+		panic(err)
+	}
+	// searching all metadata nodes and:
+	// - update store
+	// - update Nodes for ConfState
+	q := []*store.NodeExtern{}
+	q = append(q, event.Node)
+	for len(q) > 0 {
+		n := q[0]
+		q = q[1:]
+		if n.Key != etcdserver.StoreClusterPrefix {
+			v := ""
+			if !n.Dir {
+				v = *n.Value
+			}
+			if n.Key == path.Join(etcdserver.StoreClusterPrefix, "version") {
+				v = "2.3.7"
+			}
+			if _, err := st.Set(n.Key, n.Dir, v, store.TTLOptionSet{}); err != nil {
+				panic(err)
+			}
+
+			fields := strings.Split(n.Key, "/")
+			if len(fields) == 4 && fields[2] == "members" {
+				nodeID, err := strconv.ParseUint(fields[3], 16, 64)
+				if err != nil {
+					fmt.Println("wrong ID: %s", fields[3])
+					panic(err)
+				}
+				nodeIDs = append(nodeIDs, nodeID)
+			}
+		}
+		for _, next := range n.Nodes {
+			q = append(q, next)
+		}
+	}
+
+	data, err := st.Save()
+	if err != nil {
+		panic(err)
+	}
+	raftSnap := raftpb.Snapshot{
+		Data: data,
+		Metadata: raftpb.SnapshotMetadata{
+			Index: hardstate.Commit,
+			Term:  hardstate.Term,
+			ConfState: raftpb.ConfState{
+				Nodes: nodeIDs,
+			},
+		},
+	}
+	snapshotter := snap.New(path.Join(migrateDatadir, "member", "snap"))
+	if err := snapshotter.SaveSnap(raftSnap); err != nil {
+		panic(err)
+	}
+	fmt.Println("Finished.")
+}
+
+const (
+	revBytesLen            = 8 + 1 + 8
+	markedRevBytesLen      = revBytesLen + 1
+	markBytePosition       = markedRevBytesLen - 1
+	markTombstone     byte = 't'
+)
+
+func isTombstone(b []byte) bool {
+	return len(b) == markedRevBytesLen && b[markBytePosition] == markTombstone
+}
+
+func traverseAndDeleteEmptyDir(st store.Store, dir string) {
+	e, err := st.Get(dir, true, false)
+	if err != nil {
+		panic(err)
```

Comment:
I would really prefer not to have panics in functions - instead return error from it and handle it in the main function.